core/src/main/java/org/apache/iceberg/rest/HttpRESTClient.java

```diff
@@ -0,0 +1,288 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.rest.responses.ErrorResponseParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An HttpClient for usage with the REST catalog.
+ */
+public class HttpRESTClient implements RESTClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HttpRESTClient.class);
+
+  private final String uri;
+  private final CloseableHttpClient httpClient;
+  private final ObjectMapper mapper;
+  private final Map<String, String> additionalHeaders;
+  private final Consumer<HttpUriRequest> requestInterceptor;
+
+  private HttpRESTClient(
+      String uri,
+      CloseableHttpClient httpClient,
+      ObjectMapper mapper,
+      Map<String, String> additionalHeaders,
+      Consumer<HttpUriRequest> requestInterceptor) {
+    this.uri = uri;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : ImmutableMap.of();
+    this.requestInterceptor = requestInterceptor;
+  }
+
+  private static Optional<String> extractResponseBodyAsString(CloseableHttpResponse response, Class<?> responseType) {
+    try {
+      if (response.getEntity() == null) {
+        return Optional.empty();
+      }
+      return Optional.ofNullable(EntityUtils.toString(response.getEntity(), "UTF-8"));
+    } catch (IOException | ParseException e) {
+      if (responseType != null) {
+        LOG.warn("Failed to convert HTTP response body to string", e);
+      }
+      return Optional.empty();
+    }
+  }
+
+  // Per the spec, the only currently defined / used "success" responses are 200 and 202.
+  private static boolean isSuccessful(CloseableHttpResponse response) {
+    int code = response.getCode();
+    return code ==  HttpStatus.SC_OK || code == HttpStatus.SC_ACCEPTED;
+  }
+
+  private static ErrorResponse buildDefaultErrorResponse(CloseableHttpResponse response) {
+    String responseReason = response.getReasonPhrase();
+    String message =
+        responseReason != null && !responseReason.isEmpty() ? responseReason :
+            EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getCode(), null /* ignored */);
+    String type = "ClientGeneratedErrorResponse - No error response body could be parsed";
+    return ErrorResponse.builder()
+        .responseCode(response.getCode())
+        .withMessage(message)
+        .withType(type)
+        .build();
+  }
+
+  private static ErrorResponse parseOrCreateErrorResponse(
+      CloseableHttpResponse response, Optional<String> responseBody, Method method, Class<?> responseType) {
+    try {
+      return responseBody.map(ErrorResponseParser::fromJson).orElseGet(() -> buildDefaultErrorResponse(response));
+    } catch (Exception e) {
```

Comment:
There are no checked exceptions throw in the body so this would only need to catch `RuntimeException`.

But why catch an exception here and try to handle it? Why not just let the original problem propagate?

Reply:
So here, the exception that will get thrown comes from not having an `ErrorResponse` returned by the parser.

The Java optional `orElseGet` is a little confusing (at least relative to Scala). I can update this so that the `try { } catch { }` is moved out of this function so that `buildDefaultError` is still called.

Reply:
The benefit of still calling `buildDefaultError` is that it uses the response code from the response still, so the error handlers _should_ still pick up on things like ELB 5xx or client connection disconnects or something that don't even get far enough to generate a formatted `ErrorResponse`.

Reply:
I removed the use of `Optional` entirely. It was causing too many issues.

I moved the `ErrorResponseParser.fromJson` call to be higher up and added a note on why we're catching the exception and not having it propagate.